### PR TITLE
Bug fixes and refactoring code

### PR DESCRIPTION
**Pull request by MichaelS11**

### CODE REVIEW

**Comment by MichaelS11** [comment]
I hope this works better for you. Welcome any feedback.

**Comment by k0kubun** [comment]
Please separate bug fixes and refactoring into multiple commits. Also, separate each commit to any independent topics if any. Again, I have no time to review a commit including 595 lines. 100 lines are easy to understand, 200 lines are okay but it takes much time and 300 lines or over are hardly read by me except paid works.

And please describe what you did in summary in the first line of commit message and why you did refactoring or how it was wrong in the second line or later of commit message. I can't imagine the contents of the commit from "Bug fixes and refactoring code".

File: board.go [hunk 1]
```diff
@@ -1,20 +1,18 @@
...
-	"github.com/k0kubun/termbox-go"
-	"time"
+	"github.com/nsf/termbox-go"
 )
 
 const (
-	boardWidth        = 10
-	boardHeight       = 18
-	blankColor        = termbox.ColorBlack
-	animationDuration = 160
```
Comment (MichaelS11) [inline board.go]: Vars used in more than one spot moved to tetris.go
Other vars moved to file used in.

File: board.go [hunk 2]
```diff
@@ -1,20 +1,18 @@
...
+	defaultMinoX = 3
+	defaultMinoY = -1
```
Comment (MichaelS11) [inline board.go]: Used here so moved here.

File: board.go [hunk 3]
```diff
@@ -1,20 +1,18 @@
...
 )
 
 type Board struct {
-	colors [boardWidth][boardHeight]termbox.Attribute
+	colors      [boardWidth][boardHeight]termbox.Attribute
+	currentMino *Mino
+	nextMino    *Mino
 }
```
Comment (MichaelS11) [inline board.go]: Moved currentMino and nextMino into board instead of being global.

File: board.go [hunk 4]
```diff
@@ -24,6 +22,10 @@ func NewBoard() *Board {
 			board.colors[i][j] = blankColor
 		}
 	}
+	board.nextMino = NewMino()
+  board.currentMino = NewMino()
+	board.currentMino.x = defaultMinoX
+	board.currentMino.y = defaultMinoY
 	return board
```
Comment (MichaelS11) [inline board.go]: Init currentMino and nextMino in board instead of in Engine.

File: board.go [hunk 5]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
 }
 
-func (b *Board) fullLines() []int {
-	fullLines := []int{}
+func (b *Board) HasFullLine() bool {
 	for j := 0; j < boardHeight; j++ {
 		if b.isFullLine(j) {
-			fullLines = append(fullLines, j)
+			return true
 		}
```
Comment (MichaelS11) [inline board.go]: This is just from a reorder for user reading.

File: board.go [hunk 6]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
-				b.colorizeLine(line, termbox.ColorCyan)
-			}
-		})
-		timer := time.NewTimer(animationDuration * time.Millisecond)
-		<-timer.C
-
-		rewriteScreen(func() {})
-		timer = time.NewTimer(animationDuration * time.Millisecond)
-		<-timer.C
-	}
```
Comment (MichaelS11) [inline board.go]: Move animation into view

File: board.go [hunk 7]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+	return false
 }
 
-func (b *Board) colorizeLine(line int, color termbox.Attribute) {
-	for i := 0; i < boardWidth; i++ {
-		drawBack(i+boardXOffset, line+boardYOffset, color)
```
Comment (MichaelS11) [inline board.go]: Removed / moved parts into view

File: board.go [hunk 8]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+func (b *Board) FullLines() []int {
+	fullLines := []int{}
+	for j := 0; j < boardHeight; j++ {
+		if b.isFullLine(j) {
+			fullLines = append(fullLines, j)
+		}
 	}
+	return fullLines
 }
```
Comment (MichaelS11) [inline board.go]: This is just from a reorder for user reading.

File: board.go [hunk 9]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
-		}
-	}
-	return !hasBlank
-}
-
-func (b *Board) hasFullLine() bool {
-	for j := 0; j < boardHeight; j++ {
-		if b.isFullLine(j) {
-			return true
+			return false
```
Comment (MichaelS11) [inline board.go]: This is just from a reorder for user reading.

File: board.go [hunk 10]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+	return true
 }
 
-func (b *Board) text() string {
-	text := ""
-	for j := 0; j < boardHeight; j++ {
-		for i := 0; i < boardWidth; i++ {
-			text = fmt.Sprintf("%s%c", text, charByColor(b.colors[i][j]))
-		}
-		text = fmt.Sprintf("%s\n", text)
```
Comment (MichaelS11) [inline board.go]: text string no longer need, move display into view.

File: board.go [hunk 11]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+func (b *Board) setCells(cells []*Cell) {
+	for _, cell := range cells {
+		b.setCell(cell)
```
Comment (MichaelS11) [inline board.go]: No real change here

File: board.go [hunk 12]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
-	return text
 }
 
 func (b *Board) setCell(cell *Cell) {
 	b.colors[cell.x][cell.y] = cell.color
 }
 
-func (b *Board) setCells(cells []*Cell) {
-	for _, cell := range cells {
-		b.setCell(cell)
```
Comment (MichaelS11) [inline board.go]: No real change here

File: tetris.go [hunk 13]
```diff
@@ -1,149 +1,41 @@
...
-			for y := boardHeight - 1; y > boardHeight-1-j; y -= 1 {
-				board.colorizeLine(y, termbox.ColorBlack)
-			}
-		})
-		timer := time.NewTimer(gameoverDuration * time.Millisecond)
-		<-timer.C
-	}
-	clock.lock = false
-}
-
```
Comment (MichaelS11) [inline tetris.go]: Moved elsewhere, mostly to engine.

File: board.go [hunk 14]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+	b.currentMino = b.nextMino
+	b.currentMino.x = defaultMinoX
+	b.currentMino.y = defaultMinoY
+	if b.currentMino.conflicts() {
+		engine.gameOver()
+		return
 	}
+	b.nextMino = NewMino()
+}
+
```
Comment (MichaelS11) [inline board.go]: Moved from tetris.go

File: board.go [hunk 15]
```diff
@@ -41,77 +43,59 @@ func (b *Board) deleteLine(y int) {
...
+func (b *Board) ApplyGravity() {
+	board.currentMino.moveDown()
 }
```
Comment (MichaelS11) [inline board.go]: New helper

File: clock.go [hunk 16]
```diff
@@ -1,60 +0,0 @@
...
-func (c *Clock) over() {
-	c.ticker.Stop()
-	c.gameover = true
-	c.paused = true
-}
-
-func (c *Clock) pause() {
-	c.ticker.Stop()
-	c.paused = true
-}
```
Comment (MichaelS11) [inline clock.go]: Changed to engine

File: keybind.go [hunk 17]
```diff
@@ -1,48 +0,0 @@
...
-				} else if ev.Key == termbox.KeyArrowLeft {
-					currentMino.moveLeft()
-				} else if ev.Key == termbox.KeyArrowRight {
-					currentMino.moveRight()
-				}
-			}
-		}
-		refreshScreen()
-	}
-}
```
Comment (MichaelS11) [inline keybind.go]: Changed to key_input.go

File: mino.go [hunk 18]
```diff
@@ -6,64 +6,60 @@ import (
...
+MMM.
+....`,
+		`....
+.b..
+.bbb
+....`,
+		`....
+..m.
+mmm.
+....`,
```
Comment (MichaelS11) [inline mino.go]: Just flattered

File: mino.go [hunk 19]
```diff
@@ -6,64 +6,60 @@ import (
...
 	}
+	bag      []int
+	bagIndex int
 )
 
+func init() {
+	rand.Seed(time.Now().UnixNano())
+	bag = rand.Perm(7)
+}
+
```
Comment (MichaelS11) [inline mino.go]: use random bag, less bunching of pieces.

File: mino.go [hunk 20]
```diff
@@ -77,43 +73,44 @@ func (m *Mino) setCell(x, y int, cell rune) {
...
+	if !m.isOnBoard() {
+		engine.gameOver()
+		return
+	}
+
+	engine.AddScore(distance)
+	board.setCells(m.cells())
+	board.addMino()
 }
```
Comment (MichaelS11) [inline mino.go]: Combine drop and put down. Added isOnBoard check (bug fix)

File: mino.go [hunk 21]
```diff
@@ -77,43 +73,44 @@ func (m *Mino) setCell(x, y int, cell rune) {
...
+			return
+		}
 		board.setCells(m.cells())
-		pushMino()
-	} else {
-		m.forceMoveDown()
+		board.addMino()
+		return
 	}
-}
```
Comment (MichaelS11) [inline mino.go]: Added isOnBoard, bug fix

File: mino.go [hunk 22]
```diff
@@ -77,43 +73,44 @@ func (m *Mino) setCell(x, y int, cell rune) {
...
 
-func (m *Mino) forceMoveDown() {
```
Comment (MichaelS11) [inline mino.go]: Removed ForceMoveDown

File: mino.go [hunk 23]
```diff
@@ -138,6 +135,12 @@ func (m *Mino) rotateRight() {
 	dstMino.forceRotateRight()
 	if !dstMino.conflicts() {
 		m.forceRotateRight()
+		return
+	}
+	dstMino.x--
+	if !dstMino.conflicts() {
+		m.x--
+		m.forceRotateRight()
```
Comment (MichaelS11) [inline mino.go]: Added bump if near wall

File: mino.go [hunk 24]
```diff
@@ -155,6 +158,12 @@ func (m *Mino) rotateLeft() {
 	dstMino.forceRotateLeft()
 	if !dstMino.conflicts() {
 		m.forceRotateLeft()
+		return
+	}
+	dstMino.x++
+	if !dstMino.conflicts() {
+		m.x++
+		m.forceRotateLeft()
```
Comment (MichaelS11) [inline mino.go]: Added bump if near wall

File: view.go [hunk 25]
```diff
@@ -2,39 +2,9 @@ package main
...
-		WkkkkkkkkkkW
-		WkkkkkkkkkkW BBBBBB
-		WkkkkkkkkkkW WWWWWW
-		WkkkkkkkkkkW
-		WkkkkkkkkkkW
-		WWWWWWWWWWWW
-
-		kkkkkkkkkkkkkkkkkkk
-		WWWWWWWWWWWWWWWWWWW
-	`
```
Comment (MichaelS11) [inline view.go]: Changed to more dynamic code so can change board size.

File: view.go [hunk 26]
```diff
@@ -58,142 +28,242 @@ var (
...
-func refreshScreen() {
-	rewriteScreen(func() {
-		drawDropMarker()
-		drawCurrentMino()
-		if clock.gameover {
-			for j := 0; j < boardHeight; j++ {
-				board.colorizeLine(j, termbox.ColorBlack)
-			}
-			drawText(10, 4, "GAME OVER", termbox.ColorWhite, termbox.ColorBlack)
-			drawText(7, 6, "<SPC> to continue", termbox.ColorWhite, termbox.ColorBlack)
```
Comment (MichaelS11) [inline view.go]: Removed function part and put in own functions to be called.

File: view.go [hunk 27]
```diff
@@ -58,142 +28,242 @@ var (
...
+	view.drawNextMino()
+
+	if engine.gameover {
+		view.drawGameOver()
+	} else {
+		view.drawCurrentMino()
+		if !view.deleteAnimation {
+			view.drawDropMarker()
+		}
+	}
```
Comment (MichaelS11) [inline view.go]: Changed what is viewed depending on engine state or view state.

File: view.go [hunk 28]
```diff
@@ -58,142 +28,242 @@ var (
...
+			if x == xOffset || x == xOffset+1 || x == xEnd-1 || x == xEnd-2 ||
+				y == yOffset || y == yEnd-1 {
+				termbox.SetCell(x, y, ' ', termbox.ColorDefault, colorBackgroundBorder)
+			} else {
+				termbox.SetCell(x, y, ' ', termbox.ColorDefault, colorBackgroundFill)
+			}
+		}
+	}
+
+}
```
Comment (MichaelS11) [inline view.go]: More dynamic background that can change with board size.

File: view.go [hunk 29]
```diff
@@ -58,142 +28,242 @@ var (
...
+	view.drawText(xOffset, yOffset, "↓  - slow drop", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset++
+	view.drawText(xOffset, yOffset, "sbar - full drop", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset++
+	view.drawText(xOffset, yOffset, "p    - pause", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset++
+	view.drawText(xOffset, yOffset, "q    - quit", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset++
+	view.drawText(xOffset, yOffset, "u    - level up", termbox.ColorWhite, termbox.ColorBlack)
 }
```
Comment (MichaelS11) [inline view.go]: More dynamic text that can change with board size.

File: view.go [hunk 30]
```diff
@@ -58,142 +28,242 @@ var (
...
 
-func drawCell(x, y int, color termbox.Attribute) {
+func (view *View) drawCell(x, y int, color termbox.Attribute) {
 	if color != termbox.ColorDefault && color != blankColor {
-		if color == colorByChar('K') {
-			termbox.SetCell(2*x-1, y, '▓', color, termbox.ColorWhite)
-			termbox.SetCell(2*x, y, ' ', color, termbox.ColorWhite)
+		if color == colorMapping['K'] {
+			termbox.SetCell(2*x, y, '▓', color, termbox.ColorWhite)
+			termbox.SetCell(2*x+1, y, ' ', color, termbox.ColorWhite)
```
Comment (MichaelS11) [inline view.go]: Removed colorByChar and accessed color directly

File: view.go [hunk 31]
```diff
@@ -58,142 +28,242 @@ var (
...
+	view.drawText(xOffset, yOffset, "   GAME OVER", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset += 2
+	view.drawText(xOffset, yOffset, "sbar for new game", termbox.ColorWhite, termbox.ColorBlack)
+	yOffset += 2
+	xOffset += 2
+	ranking := NewRanking()
+	for index, line := range ranking.scores {
+		view.drawText(xOffset, yOffset+index, fmt.Sprintf("%2d: %6d", index+1, line), termbox.ColorWhite, termbox.ColorBlack)
 	}
 }
```
Comment (MichaelS11) [inline view.go]: Added drawGameOver function

File: view.go [hunk 32]
```diff
@@ -58,142 +28,242 @@ var (
...
 		}
+		termbox.Flush()
+		time.Sleep(160 * time.Millisecond)
+		view.RefreshScreen()
+		time.Sleep(160 * time.Millisecond)
 	}
-	return '.'
+
+	view.deleteAnimation = false
 }
```
Comment (MichaelS11) [inline view.go]: Added Show Delete Animation

**Comment by k0kubun** [comment]
While you add comments, what I said was separating "commits". Please split one commits which do multiple things to multiple commits.

**Pull request closed without merging**